File: System_model/SolverOfThreeOrderFunction.py
'''
Author: Dongming Wang wdong25@ucr.edu
Date: August 1st, 2024
'''

def SolverOfThreeOrderFunction(a, b, c, d, K):
    if d >= 1.2179 * K:
        return 1.2
    elif d == 0:
        return 0
    else:
        x1 = 0
        x2 = 1.2

        def cubic_function(x):
            return (a * x + b) * x**2 + c * x + d

        fx1 = cubic_function(x1)
        fx2 = cubic_function(x2)

        flag2 = True
        while flag2:
            x0 = (x1 + x2) / 2
            fx0 = cubic_function(x0)

            if fx0 * fx1 < 0:
                x2 = x0
                fx2 = fx0
            else:
                x1 = x0
                fx1 = fx0

            if abs(fx0) < 1e-2:
                flag2 = False

        return x0

File: Deleted/System Model/SolverOfThreeOrderFunction.m
function x0=SolverOfThreeOrderFunction(a,b,c,d,K)
if d>=1.2179*K
    x0=1.2;
elseif d==0
    x0=0;
else
x1 = 0;
x2 = 1.2;

fx1 = ((a * x1 + b) * x1 + c) * x1 + d;
fx2 = ((a * x2 + b) * x2 + c) * x2 + d;

flag2=1;
while flag2
    x0=(x1+x2)/2;
    fx0=((a*x0+b)*x0+c)*x0+d;
    
    if fx0*fx1 < 0
        x2 = x0;
        fx2 = fx0;
    else 
        x1 = x0;
        fx1 = fx0;
    end
    if abs(fx0) < 1e-2
        flag2=0;
    end
end
end
% x0=(-1*c-sqrt(c*c-4*b*d))/(4*b);
end